File: libs/parsing/src/parsing/splitter.py
"""Page-range sharding with chapter-aligned bounds (PRD §6.2).

A book is never a unit of work; a 16–24 page shard is (§1.1). Shard
boundaries snap to PDF bookmarks where available — chapters rarely
split a table or paragraph, so this eliminates most cross-boundary
chunk damage — falling back to fixed windows with a 1-page overlap the
embedder later dedupes by hash.
"""

from __future__ import annotations

from dataclasses import dataclass

from core.config import Settings, get_settings


@dataclass(frozen=True)
class ShardBound:
    idx: int
    page_start: int  # inclusive, 1-based
    page_end: int  # inclusive


def fixed_bounds(
    page_count: int,
    shard_pages: int,
    overlap: int = 1,
) -> list[ShardBound]:
    """Fixed windows with a trailing 1-page overlap between consecutive
    shards. The embedder dedupes overlapping content by chunk hash."""
    if page_count < 1:
        raise ValueError(f"page_count must be >= 1, got {page_count}")
    if shard_pages < 1:
        raise ValueError(f"shard_pages must be >= 1, got {shard_pages}")

    bounds: list[ShardBound] = []
    start = 1
    idx = 0
    while start <= page_count:
        end = min(start + shard_pages - 1, page_count)
        bounds.append(ShardBound(idx=idx, page_start=start, page_end=end))
        idx += 1
        start = end + 1 - overlap if end < page_count else end + 1
        # The overlap shifts the next start back one page, but never
        # before the current start (guard against shard_pages=1 loops).
        if start <= bounds[-1].page_start and end < page_count:
            start = bounds[-1].page_start + 1
    return bounds


def chapter_aligned_bounds(
    outline: list[tuple[int, str]],
    page_count: int,
    settings: Settings | None = None,
) -> list[ShardBound]:
    """Snap shard boundaries to bookmarks (1-based target pages).

    ``outline`` is [(page, title), ...] in document order. Between two
    consecutive chapter starts we emit fixed windows of shard_pages;
    the final window of each chapter range is allowed to run short
    rather than spill into the next chapter. A leading preamble (pages
    before the first bookmark) is sharded with fixed bounds too.
    """
    s = settings or get_settings()
    shard_pages = s.shard_pages
    clean = sorted({(max(1, min(p, page_count)), t) for p, t in outline})

    bounds: list[ShardBound] = []
    idx = 0
    prev_start = 1
    for page, _title in clean:
        # shard the range [prev_start, page-1] with fixed windows
        if page > prev_start:
            for b in fixed_bounds(page - prev_start, shard_pages, overlap=0):
                bounds.append(
                    ShardBound(
                        idx=idx,
                        page_start=b.page_start + prev_start - 1,
                        page_end=b.page_end + prev_start - 1,
                    )
                )
                idx += 1
        prev_start = page
    if prev_start <= page_count:
        for b in fixed_bounds(page_count - prev_start + 1, shard_pages, overlap=0):
            bounds.append(
                ShardBound(
                    idx=idx,
                    page_start=b.page_start + prev_start - 1,
                    page_end=b.page_end + prev_start - 1,
                )
            )
            idx += 1

    # Degenerate outline (e.g. every bookmark on page 1) can yield zero
    # or single-page shards only; fall back to plain fixed bounds.
    if not bounds or any(b.page_end < b.page_start for b in bounds):
        return fixed_bounds(page_count, shard_pages)
    return bounds


def extract_outline(pdf_path: str) -> list[tuple[int, str]]:
    """Best-effort bookmark extraction via pypdfium2; [] when absent.

    Kept tolerant: a broken outline must never fail the split stage —
    fixed bounds are always a valid fallback.
    """
    try:
        import pypdfium2 as pdfium

        pdf = pdfium.PdfDocument(pdf_path)
        try:
            out: list[tuple[int, str]] = []
            for bm in pdf.get_toc():
                if bm.page_index is not None:
                    out.append((bm.page_index + 1, bm.title or ""))
            return out
        finally:
            pdf.close()
    except Exception:
        return []

File: internal/pipeline/splitter.go
package pipeline

import (
	"fmt"
	"sort"
)

// ShardBound is one shard's 1-based inclusive page range (splitter.py).
type ShardBound struct {
	Idx       int
	PageStart int // inclusive, 1-based
	PageEnd   int // inclusive
}

// FixedBounds returns fixed windows with a trailing 1-page overlap between
// consecutive shards. The embedder dedupes overlapping content by chunk
// hash. Line-by-line port of libs/parsing splitter.py fixed_bounds —
// including the shard_pages=1 anti-loop guard (a bugfix, mandatory).
func FixedBounds(pageCount, shardPages, overlap int) ([]ShardBound, error) {
	if pageCount < 1 {
		return nil, fmt.Errorf("page_count must be >= 1, got %d", pageCount)
	}
	if shardPages < 1 {
		return nil, fmt.Errorf("shard_pages must be >= 1, got %d", shardPages)
	}
	var bounds []ShardBound
	start := 1
	idx := 0
	for start <= pageCount {
		end := start + shardPages - 1
		if end > pageCount {
			end = pageCount
		}
		bounds = append(bounds, ShardBound{Idx: idx, PageStart: start, PageEnd: end})
		idx++
		if end < pageCount {
			start = end + 1 - overlap
		} else {
			start = end + 1
		}
		// The overlap shifts the next start back one page, but never
		// before the current start (guard against shard_pages=1 loops).
		if start <= bounds[len(bounds)-1].PageStart && end < pageCount {
			start = bounds[len(bounds)-1].PageStart + 1
		}
	}
	return bounds, nil
}

// ChapterAlignedBounds snaps shard boundaries to bookmarks (1-based target
// pages). Port of splitter.py chapter_aligned_bounds, including the
// degenerate-outline fallback (a bugfix, mandatory): an outline that yields
// zero or inverted bounds falls back to plain fixed bounds.
func ChapterAlignedBounds(outline []Bookmark, pageCount, shardPages int) ([]ShardBound, error) {
	if shardPages < 1 {
		shardPages = 1
	}
	clean := map[int]string{}
	for _, bm := range outline {
		p := bm.Page
		if p < 1 {
			p = 1
		}
		if p > pageCount {
			p = pageCount
		}
		if _, ok := clean[p]; !ok || bm.Title != "" {
			clean[p] = bm.Title
		}
	}
	pages := make([]int, 0, len(clean))
	for p := range clean {
		pages = append(pages, p)
	}
	sort.Ints(pages)

	var bounds []ShardBound
	idx := 0
	prevStart := 1
	for _, page := range pages {
		// shard the range [prev_start, page-1] with fixed windows
		if page > prevStart {
			span := page - prevStart
			fb, err := FixedBounds(span, shardPages, 0)
			if err != nil {
				return nil, err
			}
			for _, b := range fb {
				bounds = append(bounds, ShardBound{
					Idx: idx, PageStart: b.PageStart + prevStart - 1,
					PageEnd: b.PageEnd + prevStart - 1,
				})
				idx++
			}
		}
		prevStart = page
	}
	if prevStart <= pageCount {
		span := pageCount - prevStart + 1
		fb, err := FixedBounds(span, shardPages, 0)
		if err != nil {
			return nil, err
		}
		for _, b := range fb {
			bounds = append(bounds, ShardBound{
				Idx: idx, PageStart: b.PageStart + prevStart - 1,
				PageEnd: b.PageEnd + prevStart - 1,
			})
			idx++
		}
	}

	// Degenerate outline (e.g. every bookmark on page 1) can yield zero
	// or single-page shards only; fall back to plain fixed bounds.
	degenerate := len(bounds) == 0
	for _, b := range bounds {
		if b.PageEnd < b.PageStart {
			degenerate = true
			break
		}
	}
	if degenerate {
		return FixedBounds(pageCount, shardPages, 1)
	}
	return bounds, nil
}

// Bookmark is one extracted PDF outline entry (1-based page).
type Bookmark struct {
	Page  int
	Title string
}
